### PR TITLE
systemd.mount integration

### PR DESCRIPTION
**Pull request by aerusso**
Implements the "systemd generator" protocol in `zfs-mount-generator`

### Description
zfs-mount-generator implements the "systemd generator" protocol, producing systemd.mount units from the (possibly cached) output of `zfs list` during early boot, giving full systemd integration.

The most visible benefit of this is that /etc/fstab can safely refer to zfs mount points, because systemd will take care to mount filesystems in the correct order.

### Motivation and Context
This PR  takes a different approach from #6974, which modified `/etc/fstab` to reflect ZFS mountpoints. Here, instead, ZFS mounts are tracked by directly creating native systemd `.mount` units, at early boot from the output of `zfs list -H -t filesystem -oname,mountpoint,canmount`. Because pools may not be imported, the output of this command can be saved in `/etc/zfs/zfs-list.cache`. If the pools are for some reason mounted at early boot (e.g., zfs on root), this file can be omitted and the command will be run.

This generator is *not required*; it does not interfere with `zfs-mount.service`, so anything missing from the cache file (or from an unimported pool) will be mounted as before.

As mentioned before, this allows for complex mount hierarchies (e.g., bind mounts that must happen after zfs mounts are made; any other filesystem mounted on top of any ZFS). Notice that ZFS on root users are most likely to want such features, and will not have to create the `zfs-list.cache` file.

### How Has This Been Tested?

I've been using several incarnations of this generator for several months, allowing for some maturity in the patches. E.g., a dependency has been reduced from `Requires` to `Wants` to prevent filesystems from being unmounted when zfs systemd units are shuffled around during upgrades.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux code style requirements.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

**Comment by Fabian-Gruenbichler** [comment]
concept LGTM at a first glance - not yet tested though ;) hopefully I'll find some time soon.

I wonder whether we want some kind of integration for re-generating the cached dataset list? it does sound a bit cumbersome to have to remember to re-run it after creating or renaming a dataset.. OTOH I am not sure how we'd be able to automate that easily.. maybe some kind of (optional) zfs-zed integration?

**Comment by AttilaFueloep** [comment]
@aerusso Thanks for that, works like a charm!

I'm running root on ZFS with homegrown boot environments and a complex fs layout (see below). This is a current Arch Linux with zfs-git kmod. Up to now all mount points were set to legacy and got mounted via `/etc/fstab`. Setting all mount points to the appropriate directories, removing them from `fstab` and rebooting gave me, as expected, only a mounted `/`.  Then I tossed in your generator and rebooted again. Now all fs got mounted properly. Nice!

I'm still seeing a `systemd[1]: var.mount: Directory /var to mount over is not empty, mounting anyway` on boot (and a `failed to unmount /var" message on shutdown`) but this was also the case with the old legacy mounts, so nothing new. Looks like `systemd` (`journald` ?) is writing to `/var` to early. This caused no havoc yet, so I can live with that.

`zfs mount` without generator:
```
rpool/ROOT/lx-4.15.10-1-ARCH_1   /
```

`zfs mount` with generator:
```
rpool/ROOT/lx-4.15.10-1-ARCH_1  /
rpool/home/root             /root
rpool/var                   /var
rpool/home                  /home
rpool/opt                   /opt
rpool/vbox                  /vbox
rpool/var/cache             /var/cache
rpool/home/me              /home/me
rpool/var/temp              /var/temp
rpool/vbox/me              /vbox/me
rpool/tmp/me               /home/me/tmp
rpool/vbox/me/vms          /vbox/me/vms
rpoolubia/home/me/stuff      /home/me/stuff
```
I don't  know much about `systemd`, but if you want me to review anyway I would happily do so. Please just drop me a note then.

Thanks again.

**Comment by aerusso** [comment]
@AttilaFueloep I'm glad it's working! Thanks for testing it. My setup is very similar, except root is not ZFS (yet). I also cannot get `/var/log` to unmount at system shutdown (but this is IMO a systemd journald bug/feature depending on how you feel about it).

As for the unclean *mounting*, I'd guess you have some residual files/directories under `/var/` that haven't been cleaned up. systemd is usually pretty careful to not *open* `/var/log/journal` until the directory is mounted (check journalctl -b, look for "Starting Flush Journal to Persistent Storage"). If indeed that's happening *after* `/var/` is mounted, you can inspect that by bind mounting `/` somewhere, and peeking inside `/var`: i.e.,

```
# mount --bind / /mnt
# cd /mnt/var
# ls
```

Then carefully remove/move anything inside there. The original motivation for this patch was the contamination of these mountpoints when things didn't get set up correctly and some services got started.

**Comment by Fabian-Gruenbichler** [comment]
On Thu, Mar 22, 2018 at 10:22:50PM +0000, Antonio Russo wrote:
> So, I took a look at `cmd/zed/zed.d/README`, and decided that I wasn't going to try to implement a new zed event (at least right now). However, I did fix up the script a bit (it works with bash and dash, instead of requiring bash). All of `shellcheck.net` valid complaints have been addressed.

see below

> 
> Also, I changed the semantics around a little bit:
> 1. The dependencies of `local-fs.target` are reduced to `Wants`. This reduces the chance of unpleasant surprises if a mount fails. Maybe it gets raised to `Requires` after it gets a little wider testing.
> 2. Instead of clobbering existing `.mount`s, abort instead.

should we maybe log this? after all, this means there is a potential
conflict between a manually set up .mount unit and a generated one
(previously generated ones are cleared before the generator is called)

> 3. `zfs list` is run unconditionally, and its output is always preferred to `/etc/zfs/zfs-list.cache`.

the last one is problematic IMHO. `zfs list` can take quite a while on a
system with lots of datasets and under load, or might (in case of bugs /
...) even hang altogether.

see `man systemd.generator`:

```
       ·   Generators are run very early at boot and cannot rely on any external services. They may not talk to any other process. That includes simple things such as logging to syslog(3), or systemd itself (this means: no
           systemctl(1))! Non-essential file systems like /var and /home are mounted after generators have run. Generators can however rely on the most basic kernel functionality to be available, including a mounted /sys, /proc,
           /dev, /usr.

[...]

       ·   If you are careful, you can implement generators in shell scripts. We do recommend C code however, since generators are executed synchronously and hence delay the entire boot if they are slow.
```

so I think any calls to external binaries, especially ones which could
block in kernel space are a no-go in generators.

I wonder whether it would not be better to drop the call to `zfs list`
altogether and invest some energy into ZED integration to keep the cache
file current? we'd basically need to hook:
- pool creation
- pool import
- pool export (or not?)
- filesystem creation (including receive and clone)
- filesystem destruction (or not?)
- filesystem rename
- filesystem mountpoint property changes
- filesystem canmount property changes

not sure which of those are already available in ZED? if you don't want
to do this I can try to whip something up..

> 
> This should further reduce the need to populate `zfs-list.cache` (in fact, just running `systemctl daemon-reload` will re-run generators, producing the desired mount units).
> 

see above for why the fact that systemd will re-run generators on every
reload makes matters worse, not better ;)

e.g. a single upgrade on a Debian testing or unstable might run
`systemctl daemon-reload` tens or hundreds of times (depending on the
number of upgraded packages which contain unit files of some kind)!

I am not sure how the RPM world handles upgraded unit files..

on my system with ~256 datasets, running the generator already takes a
measurable fraction of a second, but I guess most of that is actually
writing the unit files (haven't checked yet)

> It would be nice if someone running zfs on root could give this script a try (just toss it in `/etc/system/system-generators` after replacing `@sysconfdir@` and `@sbindir@` with the correct values).

FWIW, I did on Stretch and Sid. it works fine, and allows me to drop a
`Requires` on zfs-mount.service from a bind mount :) so I'd very much
like to see this integrated in some form!

the testing instructions lack a `chmod +x
/etc/systemd/system-generators/zfs-mount-generator` though ;)

**Comment by aerusso** [comment]
I've reworked this again using @Fabian-Gruenbichler's suggestions:

>> 2. Instead of clobbering existing `.mount`s, abort instead.
>
>should we maybe log this? after all, this means there is a potential
>conflict between a manually set up .mount unit and a generated one
>(previously generated ones are cleared before the generator is called)

Done.

> I wonder whether it would not be better to drop the call to `zfs list`
> altogether and invest some energy into ZED integration to keep the cache
> file current? we'd basically need to hook:
> - pool creation
> - pool import
> - pool export (or not?)
> - filesystem creation (including receive and clone)
> - filesystem destruction (or not?)
> - filesystem rename
> - filesystem mountpoint property changes
> - filesystem canmount property changes

No call to `zfs list` is made in the systemd-generator anymore---if there's any concern over long hangs or serious bugs, that absolutely must not be allowed to interfere with system startup.

The new patch implements a `history_event-zfs-list-cacher.sh` "ZEDLET" updating `zfs-list.cache`. It tracks `destroy`, `rename`, `mountpoint` and `canmount` changes. I don't ever add new ones automatically, keeping the administrator in control (**How do we feel about this?**).

File: cmd/zed/zed.d/history_event-zfs-list-cacher.sh.in [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    exit 0
+   ;;
+esac
+
+zed_lock zfs-list
+trap abort_alter EXIT
+
+# Get the information for the affected pool
+grep "^${ZEVENT_POOL}[/$(printf '\t')]" "${FSLIST}" |
+grep -o '[^'"$(printf '\t')"'*' |
```
Comment (Fabian-Gruenbichler) [inline cmd/zed/zed.d/history_event-zfs-list-cacher.sh.in]: `grep: Unmatched [ or [^`

I think the intended semantics was that of `cut -f 1`?

File: cmd/zed/zed.d/history_event-zfs-list-cacher.sh.in [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  xargs @sbindir@/zfs list -H -oname,mountpoint,canmount \
```
Comment (Fabian-Gruenbichler) [inline cmd/zed/zed.d/history_event-zfs-list-cacher.sh.in]: `@sbindir@/zfs` should be `${ZFS}`

File: cmd/zed/zed.d/history_event-zfs-list-cacher.sh.in [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    >"${FSLIST_TMP}" 2>/dev/null || true
```
Comment (Fabian-Gruenbichler) [inline cmd/zed/zed.d/history_event-zfs-list-cacher.sh.in]: depending on whether we want to react to any changes (see bigger comment above), the whole preceding section could be replaced with
`${ZFS} list -H -t filesystem -oname,mountpoint,canmount -r ${ZEVENT_POOL} > ${FSLIST_TMP}`

**Comment by aerusso** [comment]
I'm possibly over-engineering this patch, but I've included the ability to specify regular expressions for datasets that should be automatically added. I.e., if you want everything you import, create, or receive to be tracked with this, you can put a line
```
+ .*
```
If you just want everything in some pool, say `bigpool`
```
+ bigpool/.*
```
Also, renamed datasets should now be correctly followed, and `inherit` events are now also tracked (since they can change mountpoints). Note that renamed datasets remain tracked, even if they no longer match the regular expression that caused them to be tracked initially. Exported pools have their entries removed from the cache.

**Comment by Fabian-Gruenbichler** [comment]
ha, yes, that looks a bit over-engineered to me as well ;) how about the following:

1. regenerate cache from scratch based on currently available filesystems, no matter which event triggered
2. filter for those matched by a `+ ...` expression
3. remove those matched by a `- ...` expression

that would end up being a lot less code for almost the same result?

I wonder whether moving the expressions to two variables in zed.rc (whitelist and blacklist) and implying a whitelist of `.*` unless explicitly specified might not make matters easier? I guess the main use case is to exclude certain subtrees of a pool (e.g. where backups are received or cloned to).

minor unrelated nit: most other zedlets use 4 spaces for indentation, it probably makes sense to follow that style. the case statements are currently also pretty weirdly indented in parts..

**Comment by Fabian-Gruenbichler** [comment]
if we want to keep the current "try to update changed parts of existing cache file" approach, I would highly recommend moving to something other than shell (then we could just map the existing cache file to some hash/dict structure, update that as needed, and write it back out). not sure whether that is acceptable given that all the current zedlets are shell scripts?

maybe @behlendorf can chime in on that point?

**Comment by behlendorf** [comment]
The ZEDLET infrastructure was originally designed so they could be any manor of executable.  However, the hope was that the majority of them would be simple enough that they could be written in shell so they'd be trivial to inspect and customize.  These are getting a little on the long side but I don't think they're too unweildy yet!

**Comment by aerusso** [comment]
Here's another stab.

1. It doesn't try to track changes. It just runs `zfs list` on the affected *pool*.
2. The "cache" files are really just a cache---there's no configuration information hiding inside the file. This was bothering me, and I think @Fabian-Gruenbichler was honing in on this, too. 
3. It's much simpler. It doesn't parse the output at all.

Also, I've tested it out with a broken entry in the cache. It, as expected, doesn't interfere with the boot more than just showing that a unit failed (because the dependency is at `Wants` rather than `Requires`).  I honestly can't really think of any good reason to disable this on any datasets, so I think the complicated policies allowing administrators to disable this feature at a very fine level are a waste.

**Comment by aerusso** [comment]
@behlendorf I'm happy with it, and the biggest weakness is the lack of testing, though I have it running on several machines, and the systemd generator seems robust on them. We can always enable the zedlet later, after wider testing.

**Comment by Fabian-Gruenbichler** [comment]
On Tue, Apr 03, 2018 at 11:51:12PM +0000, Brian Behlendorf wrote:
> behlendorf commented on this pull request.
> 
> I haven't had a chance to test this, I've only read through the PR but the approach looks reasonable.  @Fabian-Gruenbichler do you have any additional concerns?
> 
> @aerusso how would you like to proceed.  Since the zedlet isn't enabled by default merging this is low risk.  That woud make it a little easier to get some additional testing on other distributions.

I still think some way to disable generation of the mount unit for
certain datasets would be good to have for the zedlet. think of the
following scenario:

tank/baz/bar  /bar  noauto
tank/foo/bar  /bar  on

depending on the order, the noauto one will be generated, but the auto
one will not! arguably we could say that you need to set canmount to off
in such a case (then this should be documented as a limitation of the
zedlet?) - but having two optional variables for whitelisting and
blacklisting the datasets (which just get mapped to two grep calls)
would IMHO be a nice followup to handle received or cloned datasets more
easily.

no blocker for me though - if you want I could handle that as a
separate PR?

**Comment by aerusso** [comment]
@Fabian-Gruenbichler That's a good point. This revision has the generator ignore datasets with `canmount=noauto`, which should make this work out-of-the-box (and I'm not even sure why we would want to make systemd mount units for `canmount=noauto` datasets).

**Comment by Fabian-Gruenbichler** [comment]
On Wed, Apr 04, 2018 at 11:56:19AM +0000, Antonio Russo wrote:
> @Fabian-Gruenbichler That's a good point. This revision has the generator ignore datasets with `canmount=noauto`, which should make this work out-of-the-box (and I'm not even sure why we would want to make systemd mount units for `canmount=noauto` datasets).
> 

sorry, missed that change - I think assuming only one dataset with
canmount=on for each mountpoint path is a reasonable choice for now.

we could also do two passes, one with only 'on', one with only 'noauto'
to implicitly order them. but all of that is (potential) follow-up
material IMHO.

File: etc/systemd/system-generators/zfs-mount-generator.in [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  mountfile="$(systemd-escape "${2#?}").mount"
+
+  # If the mountpoint has already been created, give it precedence.
+  if [ -e "${dest_norm}/${mountfile}" ] ; then
+    printf 'zfs-mount-generator.sh: %s.mount already exists\n' "${2}" \
+      >/dev/kmsg
+    return
+  fi
+
+  # heredocs may not work in very early boot, so we'll use printf instead.
```
Comment (rlaager) [inline etc/systemd/system-generators/zfs-mount-generator.in]: This "may" concerns me. Are they confirmed to not work? I can't think of why they wouldn't, but I haven't tested either.

Comment (aerusso) [reply etc/systemd/system-generators/zfs-mount-generator.in]: I cannot reproduce this issue anymore, and it's really weird so I'm just going to use a heredoc.

**Comment by aerusso** [comment]
@behlendorf I'm not yet using any zfs encryption, so I'm probably not the best person to do this. But, it looks to me like we would need another systemd unit to perform the `zfs load-key` (I don't think `Service` sections, or arbitrary executions, are allowed in mount units).

Maybe the most straightforward place to put it would probably be right after `zpool import`? I'm kind of inclined to have this in another PR, because it seems like it deserves at least a little bit of conversation.

Also, I added the last change as a normal commit to make looking at the differences easier. Obviously, that should be squashed when it's merged.

**Comment by behlendorf** [comment]
Agreed, a new PR is the way to go for adding what's needed for encrypted datasets.

Then this PR is ready to be merged, we'll follow up with new PRs as appropriate for additional improvements.  Thanks!

**Comment by aerusso** [comment]
@AttilaFueloep Hey sorry that your boot broke, but I'm glad you were able to fix it. The format of `/etc/zfs/zfs-list.cache` (which is now a directory) is kind of fluctuating a bit. #7453 is going to change the format a little bit more. I did not think about backwards compatibility. 

Remove the `/etc/zfs/zfs-list.cache` file and creating a directory with the same name. `touch` the name of the pool which you want tracked. This signals to the zed script to track that pool. Then, do anything that changes a `mountpoint` or `canmount` on that pool (forcing the script to update it). You can confirm that the script is working by running `systemctl daemon-restart` and inspecting the mount files in `/run/systemd/generator`.

I'm traveling, so I may be less available. Hope this works for you!

**Comment by AttilaFueloep** [comment]
@aerusso 
> Hey sorry that your boot broke

Don't worry, that's the risk I'm taking running git master and I don't think there's any reason worrying about backwards compatibility either, given that only two people tried out the old version.

The strange thing is that I didn't had the `/etc/zfs/zfs-list.cache` file but still the mount units were created during boot using the old generator.  I followed your advice but it turns out that I had to do an additional step: symlinking  `history_event-zfs-list-cacher.sh`  into `/etc/zfs/zed.d`. After doing that everything is working again. This is quite obvious and I fell silly not doing it earlier.

Thanks a lot for your help!

**Comment by aerusso** [comment]
@AttilaFueloep  An earlier version of this PR enumerated the datasets during system boot, rather than from the cache. It was argued that `zfs list` may be slow or unreliable on some systems (and furthermore, it is not universally available). It was dropped.

Just to reiterate, there may also be another format change in that PR, so you may need to regenerate those cache files again.

**Comment by AttilaFueloep** [comment]
@aerusso  Thanks for your explanations. I really feel bad for not reading the whole thread thoroughly before posting, everything was there. To my excuse, maybe my broken system made me a bit to nervous.

I've already added a toggle of the canmount property of a test filesystem on the root pool to my boot environment script. If I understand correctly this should trigger the zed script to update the cache file. On a side note, while this works for updating the active BE, it won't for updating a non-active BE and I've no easy solution here. What would you think of making `history_event-zfs-list-cacher.sh` callable outside of zed?  Maybe something along the lines of [this](https://gist.github.com/AttilaFueloep/958a746ee623a9e58a9fd2c95bbe370f)?

**Comment by rlaager** [comment]
Have you considered moving the cache file to /boot (which is presumably shared across BEs) and symlinking. That way, when you update it, the non-active BE also sees the changes?

If /boot is not shared across BEs in your setup, then I don’t have a great idea either. Though this may not matter, e.g. if the initramfs is doing the mounting anyway.

**Comment by AttilaFueloep** [comment]
In my setup /boot is on the zfs root filesystem and therefore not shared. I have separate filesystems for eg. /home and /var and initramfs (initcpio) does not mount them. Before this PR I had all mountpoints set to legacy and handled mounting via /etc/fstab, which is quite inconvenient. So this PR is a substantial improvement. If I could call `history_event-zfs-list-cacher.sh` directly I could script something like `/newbe/etc/zfs/zed.d/history_event-zfs-list-cacher.sh rpool >/newbe/etc/zfs/zfs-list.cache/rpool`. That would be more than enough.

**Comment by ccope** [comment]
Should this work on Ubuntu 16.04? I've copied this zedlet into /etc/zfs/zed.d and chown+chmod to root:root 755), but it doesn't fire when I run zpool import/export. Verbose output from zed:
```
Apr 30 16:47:50 htpc-nuc zed[11655]: Registered zedlet "history_event-zfs-list-cacher.sh"
Apr 30 16:47:50 htpc-nuc zed[11655]: Registered zedlet "resilver.finish-notify.sh"
Apr 30 16:47:50 htpc-nuc zed[11655]: Registered zedlet "scrub.finish-notify.sh"
Apr 30 16:47:50 htpc-nuc zed[11655]: Ignoring "zed.rc": not executable by user
Apr 30 16:47:50 htpc-nuc zed[11655]: Registered zedlet "all-syslog.sh"
Apr 30 16:47:50 htpc-nuc zed[11655]: Registered zedlet "io-notify.sh"
Apr 30 16:47:50 htpc-nuc zed[11655]: Registered zedlet "io-spare.sh"
Apr 30 16:47:50 htpc-nuc zed[11655]: Registered zedlet "data-notify.sh"
Apr 30 16:47:50 htpc-nuc zed[11655]: Registered zedlet "checksum-spare.sh"
Apr 30 16:47:50 htpc-nuc zed[11655]: Ignoring "zed-functions.sh": not executable by user
Apr 30 16:47:50 htpc-nuc zed[11655]: Registered zedlet "checksum-notify.sh"
Apr 30 16:47:50 htpc-nuc zed[11655]: ZFS Event Daemon 0.6.5.6-0ubuntu20 (PID 11655)
Apr 30 16:47:50 htpc-nuc zed[11655]: Processing events since eid=22
Apr 30 16:48:14 htpc-nuc zed[11655]: Invoking "all-syslog.sh" eid=23 pid=11715
Apr 30 16:48:14 htpc-nuc zed[11655]: Finished "all-syslog.sh" eid=23 pid=11715 exit=0
```

**Comment by aerusso** [comment]
You have to `touch /etc/zfs/zfs-list.cache/POOLNAME`  to enable this for a given pool. Did you also grab the `zfs-mount-generator` and put it in the right systemd directory?

Notice also that #7453 will change these formats relatively soon.

**Comment by ccope** [comment]
Ah, I missed the "touch" step. I generated the cache file by running the zfs list command by hand and then the generator started working. Thanks for the heads up, I'll subscribe to that PR.

File: etc/systemd/system-generators/zfs-mount-generator.in [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+set -ef
+
+FSLIST="@sysconfdir@/zfs/zfs-list.cache"
+
+[ -d "${FSLIST}" ] || exit 0
+
+do_fail() {
+  printf 'zfs-mount-generator.sh: %s\n' "$*" > /dev/kmsg
```
Comment (AttilaFueloep) [inline etc/systemd/system-generators/zfs-mount-generator.in]: @aerusso Shouldn't that read `zfs-mount-generator` instead of `zfs-mount-generator.sh`?

Comment (aerusso) [reply etc/systemd/system-generators/zfs-mount-generator.in]: Good catch. I've got a queue of documentation typos, I'll add these to it.

Comment (AttilaFueloep) [reply etc/systemd/system-generators/zfs-mount-generator.in]: Yeah, it's just cosmetic. I now tend to use `myname=$(basename "$0")` in my scripts after tripping over it as well.

**Comment by rugubara** [comment]
May I suggest a minor edit to the man page?
The suggested command misses -r and -t filesystem options. If the pools have zvols, generator fails with invalid canmount. So the correct initialization command that worked for me:
`zfs list -t filesystem -r -H -oname,mountpoint,canmount`

The generator code doesn't provide any hint in the message when this fails. I suggest to replace 
`do_fail "invalid canmount"` with
`do_fail "invalid canmount ${1} ${2} ${3}"`

**Comment by rugubara** [comment]
May I suggest a minor edit to the man page?
The suggested command misses -r and -t filesystem options. If the pools have zvols, generator fails with invalid canmount. So the correct initialization command that worked for me:
`zfs list -t filesystem -r -H -oname,mountpoint,canmount`

The generator code doesn't provide any hint in the message when this fails. I suggest to replace 
`do_fail "invalid canmount"` with
`do_fail "invalid canmount ${1} ${2} ${3}"`

**Comment by rugubara** [comment]
May I suggest a minor edit to the man page?
The suggested command misses -r and -t filesystem options. If the pools have zvols, generator fails with invalid canmount. So the correct initialization command that worked for me:
`zfs list -t filesystem -r -H -oname,mountpoint,canmount`

The generator code doesn't provide any hint in the message when this fails. I suggest to replace 
`do_fail "invalid canmount"` with
`do_fail "invalid canmount ${1} ${2} ${3}"`

**Comment by aerusso** [comment]
@rugubara How did a zvol wind up in the zfs-list cache? That's a bug in the ZED-let.

**Comment by rugubara** [comment]
@aerusso, I couldn't get zedlet working for me yet. I wrote /etc/zfs/zfs-list.cache/pool myself using the zfs list command from man page.

**Comment by aerusso** [comment]
Thank you for confirming that I'm not the only one who cannot see all the replies!

The man page has a little, tricky, caveat:

> for datasets that should be mounted by systemd

though the real bug is that the zedlet isn't working for you. What happened there?

**Comment by rugubara** [comment]
@aerusso, I forgot to symlink history_event-zfs-list-cacher.sh to /etc/zfs/zed.d. Once symlinked, everything works ok.

**Comment by mskarbek** [comment]
Is this something that could go live with 0.7.12 or do we have to wait for 0.8.0?

**Comment by AttilaFueloep** [comment]
@aerusso 
> I also cannot get `/var/log` to unmount at system shutdown (but this is IMO a systemd journald bug/feature depending on how you feel about it).

In case you still have the unmounting problem, the solution is described in https://github.com/zfsonlinux/zfs/issues/8060#issuecomment-433017531. That fixed it for me.

**Pull request merged**